client/invoke.go

```diff
@@ -0,0 +1,109 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"os"
+
+	"github.com/fnproject/fn_go/provider"
+	"github.com/go-openapi/runtime/logger"
+)
+
+func Invoke(provider provider.Provider, appName, fnID string, content io.Reader, output io.Writer, method string, env []string, contentType string, includeCallID bool) error {
+
+	u, err := provider.CallURL(appName)
+	if err != nil {
+		return err
+	}
+	u.Path = "invoke/" + fnID
+	method = "POST"
+
+	// Read the request body (up to the maximum size), as this is used in the
+	// authentication signature
+	var req *http.Request
+	if content != nil {
+		b, err := ioutil.ReadAll(io.LimitReader(content, MaximumRequestBodySize))
+		buffer := bytes.NewBuffer(b)
+		req, err = http.NewRequest(method, u.String(), buffer)
+		if err != nil {
+			return fmt.Errorf("Error creating request to service: %s", err)
+		}
+	} else {
+		var err error
+		req, err = http.NewRequest(method, u.String(), nil)
+		if err != nil {
+			return fmt.Errorf("Error creating request to service: %s", err)
+		}
+	}
+
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	} else {
+		req.Header.Set("Content-Type", "text/plain")
+	}
+
+	if len(env) > 0 {
+		EnvAsHeader(req, env)
+	}
+
+	transport := provider.WrapCallTransport(http.DefaultTransport)
+	httpClient := http.Client{Transport: transport}
+
+	if logger.DebugEnabled() {
+		b, err := httputil.DumpRequestOut(req, content != nil)
+		if err != nil {
+			return err
+		}
+		fmt.Printf(string(b) + "\n")
+	}
+
+	resp, err := httpClient.Do(req)
```

Comment:
You seem to use `resp` here before checking `err`. I think there's a risk of a segfault.

Reply:
moved the err check